File: app/src/main/java/org/softeg/slartus/forpdaplus/controls/ExtEditTextPreference.java
package org.softeg.slartus.forpdaplus.controls;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.TypedArray;
import android.os.Build;
import android.preference.EditTextPreference;
import android.util.AttributeSet;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.RequiresApi;

import org.softeg.slartus.forpdacommon.StringUtilsKt;
import org.softeg.slartus.forpdaplus.App;
import org.softeg.slartus.forpdaplus.R;

import static org.softeg.slartus.forpdaplus.utils.Utils.getS;

/*
 * Created by slinkin on 15.08.13.
 */
public class ExtEditTextPreference extends EditTextPreference {
    private enum InputType {
        Number,
        NumberDecimal
    }

    private InputType m_InputType = InputType.Number;
    private String m_DefaultValue = null;
    private CharSequence m_DefaultSummary = null;

    public ExtEditTextPreference(Context context) {
        super(context);
    }

    public ExtEditTextPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public ExtEditTextPreference(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context, attrs);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public ExtEditTextPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.ExtEditTextPreference);
        try {
            int i = a.getInt(R.styleable.ExtEditTextPreference_myInputType, -1);
            switch (i) {
                case 0:
                    m_InputType = InputType.Number;

                    break;
                case 1:
                    m_InputType = InputType.NumberDecimal;
                    break;
            }

            m_DefaultValue = a.getString(R.styleable.ExtEditTextPreference_appDefaultValue);
        } finally {
            a.recycle();
        }

        m_DefaultSummary = getSummary();
        setCurrentSummary();
    }

    @Override
    public String getText() {
        return getPersistedString(m_DefaultValue);
    }

    @Override
    protected String getPersistedString(String defaultReturnValue) {
        defaultReturnValue = m_DefaultValue;
        String result = defaultReturnValue;
        try {
            result = super.getPersistedString(defaultReturnValue);
        } catch (Throwable ex) {
            switch (m_InputType) {
                case Number:
                    result = String.valueOf(getPersistedInt(defaultReturnValue == null ? 0 : Integer.parseInt(defaultReturnValue)));
                    break;
                case NumberDecimal:
                    result = String.valueOf(getPersistedFloat(defaultReturnValue == null ? 0 : Float.parseFloat(defaultReturnValue)));
                    break;
            }
        }

        return StringUtilsKt.simplifyNumber(result);
    }

    @Override
    protected float getPersistedFloat(float defaultReturnValue) {
        try {
            return super.getPersistedFloat(defaultReturnValue);
        } catch (Throwable ex) {

            return Float.parseFloat(getPersistedString(Float.toString(defaultReturnValue)));

        }
    }

    @Override
    protected boolean persistFloat(float value) {
        if (shouldPersist()) {
            if (value == getPersistedFloat(Float.NaN)) {
                // It's already there, so the same as persisting
                return true;
            }

            SharedPreferences.Editor editor = getPreferenceManager().getSharedPreferences().edit();
            editor.putFloat(getKey(), value);
            editor.apply();
            return true;
        }
        return false;
    }

    @Override
    protected boolean persistString(String value) {
        try {
            switch (m_InputType) {
                case Number:
                    return persistInt(Integer.parseInt(value));
                case NumberDecimal:
                    float fvalue = Float.parseFloat(value);
                    return persistFloat(fvalue);
            }

            return true;
        } catch (Throwable ex) {
            Toast.makeText(getContext(), R.string.invalid_number_format, Toast.LENGTH_SHORT).show();
        }
        return false;

    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        super.onDialogClosed(positiveResult);
        setCurrentSummary();
    }

    private void setCurrentSummary() {
        String value = "";
        switch (m_InputType) {
            case Number:
                value = Integer.toString(App.getInstance().getPreferences().getInt(getKey(), Integer.parseInt(m_DefaultValue)));
                break;
            case NumberDecimal:
                value = Float.toString(App.getInstance().getPreferences().getFloat(getKey(), Float.parseFloat(m_DefaultValue)));
                break;
        }

        setSummary(String.format(m_DefaultSummary.toString(), StringUtilsKt.simplifyNumber(value)));
    }
}

File: app/src/main/java/org/softeg/slartus/forpdaplus/controls/ExtEditTextPreference.kt
package org.softeg.slartus.forpdaplus.controls

import android.content.Context
import org.softeg.slartus.forpdacommon.simplifyNumber
import androidx.annotation.RequiresApi
import android.os.Build
import android.content.res.TypedArray
import org.softeg.slartus.forpdaplus.R
import android.content.SharedPreferences
import android.util.AttributeSet
import android.widget.Toast
import androidx.preference.EditTextPreference
import org.softeg.slartus.forpdaplus.App

/*
 * Created by slinkin on 15.08.13.
 */
class ExtEditTextPreference : EditTextPreference {
    private enum class InputType {
        Number, NumberDecimal
    }

    private var m_InputType = InputType.Number
    private var m_DefaultValue: String? = null
    private var m_DefaultSummary: CharSequence? = null

    constructor(context: Context?) : super(context) {}
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(context, attrs)
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(context, attrs)
    }

    private fun init(context: Context, attrs: AttributeSet) {
        val a = context.obtainStyledAttributes(attrs, R.styleable.ExtEditTextPreference)
        try {
            val i = a.getInt(R.styleable.ExtEditTextPreference_myInputType, -1)
            when (i) {
                0 -> m_InputType = InputType.Number
                1 -> m_InputType = InputType.NumberDecimal
            }
            m_DefaultValue = a.getString(R.styleable.ExtEditTextPreference_appDefaultValue)
        } finally {
            a.recycle()
        }
        m_DefaultSummary = summary
        setCurrentSummary()
    }

    override fun getText(): String {
        return getPersistedString(m_DefaultValue!!)
    }

    override fun getPersistedString(defaultReturnValue: String): String {
        var defaultReturnValue: String? = defaultReturnValue
        defaultReturnValue = m_DefaultValue
        var result = defaultReturnValue
        result = try {
            super.getPersistedString(defaultReturnValue)
        } catch (ex: Throwable) {
            when (m_InputType) {
                InputType.Number -> getPersistedInt(defaultReturnValue?.toInt() ?: 0).toString()
                InputType.NumberDecimal -> getPersistedFloat(defaultReturnValue?.toFloat()
                        ?: 0f).toString()
            }
        }
        return result!!.simplifyNumber()
    }

    override fun getPersistedFloat(defaultReturnValue: Float): Float {
        return try {
            super.getPersistedFloat(defaultReturnValue)
        } catch (ex: Throwable) {
            getPersistedString(java.lang.Float.toString(defaultReturnValue)).toFloat()
        }
    }

    override fun persistFloat(value: Float): Boolean {
        if (shouldPersist()) {
            if (value == getPersistedFloat(Float.NaN)) {
                // It's already there, so the same as persisting
                return true
            }
            val editor = preferenceManager.sharedPreferences.edit()
            editor.putFloat(key, value)
            editor.apply()
            return true
        }
        return false
    }

    override fun persistString(value: String): Boolean {
        try {
            return when (m_InputType) {
                InputType.Number -> persistInt(value.toInt())
                InputType.NumberDecimal -> {
                    val fvalue = value.toFloat()
                    persistFloat(fvalue)
                }
            }
            return true
        } catch (ex: Throwable) {
            Toast.makeText(context, R.string.invalid_number_format, Toast.LENGTH_SHORT).show()
        }
        return false
    }


    private fun setOwnSummary(summary: CharSequence) {
        if (getSummary() != summary) {
            setSummary(summary)
        }
    }

    override fun setText(text: String?) {
        val prevValue = getText()
        super.setText(text)
        if (text != prevValue)
            setCurrentSummary()
    }
    private fun setCurrentSummary() {
        var value = ""
        value = when (m_InputType) {
            InputType.Number -> Integer.toString(App.getInstance().preferences.getInt(key, m_DefaultValue!!.toInt()))
            InputType.NumberDecimal -> java.lang.Float.toString(App.getInstance().preferences.getFloat(key, m_DefaultValue!!.toFloat()))
        }
        setOwnSummary(String.format(m_DefaultSummary.toString(), value.simplifyNumber()))
    }
}